{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accessing the Allen Brain Institute _In Situ_ Hybridization Data\n",
    "\n",
    "The Allen Brain Atlas API is used to pull differential gene expression from the mouse _in situ_ hybridization (ISH) data. This compares a set of target structures to a set of contrast structures and gives a **fold-change** value that reveals the expression of reported genes in the target structures relative to the contrast structures.\n",
    "\n",
    "Here, the target structures are layers 2/3 of the mouse visual cortex, and the contrast structures the layer 5 visual cortex areas. The data pulled from the API is then converted into a Pandas Dataframe so that it can more easily be viewed and analyzed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1. Setup.\n",
    "\n",
    "First, we need to import two different packages to communicate with the API: `requests` and `json`. As usual, we'll also import `pandas` as `pd`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2. Configure API query.\n",
    "\n",
    "Below, we'll define the `service`, or URL, where we can query the database. This will vary slightly based on the kind of search you'd like to conduct. See the Allen Brain Institute documentation on [RESTful Model Access](http://help.brain-map.org/pages/viewpage.action?pageId=5308449) for more details.\n",
    "\n",
    "We can then define `structures` that we'd like to compare. We are also going to instruct the query to use the 'Mouse Brain Atlas' ontology.\n",
    "\n",
    "We'll first query the API to find the structure IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "not enough values to unpack (expected 2, got 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-78142d193f78>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;31m# Create a single string for each set of structure IDs to use in API\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0mtemp1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtemp2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mids_1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0mtemp1\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mi\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m','\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: not enough values to unpack (expected 2, got 0)"
     ]
    }
   ],
   "source": [
    "service = 'http://api.brain-map.org/api/v2/data/query.json?criteria=' \n",
    "structures_1 = ['Primary visual area, layer 2/3']\n",
    "structures_2 = ['Primary visual area, layer 5']\n",
    "ontology = 'Mouse Brain Atlas'\n",
    "\n",
    "# Get IDs for examined structures\n",
    "ids = []\n",
    "\n",
    "for item in structures_1:\n",
    "    structure_id = requests.get('%smodel::Structure,\\\n",
    "        rma::criteria,[name$il\\'%s\\'],ontology[name$eq\\'%s\\']' % (service,item,ontology)).json()['msg'][0]['id']\n",
    "    ids_1.append(str(structure_id))\n",
    "    \n",
    "for item in structures_2:\n",
    "    structure_id = requests.get('%smodel::Structure,\\\n",
    "        rma::criteria,[name$il\\'%s\\'],ontology[name$eq\\'%s\\']' % (service,item,ontology)).json()['msg'][0]['id']\n",
    "    ids_2.append(str(structure_id))\n",
    "    \n",
    "# Create a single string for each set of structure IDs to use in API\n",
    "temp1 = ''\n",
    "temp2 = ''\n",
    "for i in ids_1:\n",
    "    temp1 += i + ','\n",
    "for i in ids_2:\n",
    "    temp2 += i + ','\n",
    "structures1 = temp1[0:-1]\n",
    "structures2 = temp2[0:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3. Run query & organize results.\n",
    "Now that we've configured our search, we can query the \"mouse_differential\" service to get differential results for our two brain structures. We'll then put the results into a Pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>gene-id</th>\n",
       "      <th>gene-symbol</th>\n",
       "      <th>gene-name</th>\n",
       "      <th>entrez-id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>plane-of-section</th>\n",
       "      <th>specimen-id</th>\n",
       "      <th>fold-change</th>\n",
       "      <th>target-sum</th>\n",
       "      <th>contrast-sum</th>\n",
       "      <th>num-target-samples</th>\n",
       "      <th>num-contrast-samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100055069</td>\n",
       "      <td>RP_080110_01_A05</td>\n",
       "      <td>11714</td>\n",
       "      <td>Atp1a1</td>\n",
       "      <td>ATPase, Na+/K+ transporting, alpha 1 polypeptide</td>\n",
       "      <td>11928</td>\n",
       "      <td>3</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>n/a</td>\n",
       "      <td>10.208</td>\n",
       "      <td>1586.732</td>\n",
       "      <td>171.975</td>\n",
       "      <td>94</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>73423932</td>\n",
       "      <td>RP_051019_03_B09</td>\n",
       "      <td>18312</td>\n",
       "      <td>Pcp4</td>\n",
       "      <td>Purkinje cell protein 4</td>\n",
       "      <td>18546</td>\n",
       "      <td>16</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>n/a</td>\n",
       "      <td>8.739</td>\n",
       "      <td>1801.084</td>\n",
       "      <td>206.098</td>\n",
       "      <td>106</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>545100</td>\n",
       "      <td>RP_040825_02_D01</td>\n",
       "      <td>46815</td>\n",
       "      <td>Pgm2l1</td>\n",
       "      <td>phosphoglucomutase 2-like 1</td>\n",
       "      <td>70974</td>\n",
       "      <td>7</td>\n",
       "      <td>sagittal</td>\n",
       "      <td>n/a</td>\n",
       "      <td>7.103</td>\n",
       "      <td>1802.017</td>\n",
       "      <td>284.020</td>\n",
       "      <td>92</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>112202624</td>\n",
       "      <td>RP_110421_01_F02</td>\n",
       "      <td>60343</td>\n",
       "      <td>Klf7</td>\n",
       "      <td>Kruppel-like factor 7 (ubiquitous)</td>\n",
       "      <td>93691</td>\n",
       "      <td>1</td>\n",
       "      <td>coronal</td>\n",
       "      <td>n/a</td>\n",
       "      <td>6.891</td>\n",
       "      <td>920.788</td>\n",
       "      <td>151.672</td>\n",
       "      <td>74</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>772</td>\n",
       "      <td>RP_Baylor_102947</td>\n",
       "      <td>18312</td>\n",
       "      <td>Pcp4</td>\n",
       "      <td>Purkinje cell protein 4</td>\n",
       "      <td>18546</td>\n",
       "      <td>16</td>\n",
       "      <td>coronal</td>\n",
       "      <td>n/a</td>\n",
       "      <td>6.556</td>\n",
       "      <td>1130.932</td>\n",
       "      <td>194.852</td>\n",
       "      <td>108</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id              name  gene-id gene-symbol  \\\n",
       "0  100055069  RP_080110_01_A05    11714      Atp1a1   \n",
       "1   73423932  RP_051019_03_B09    18312        Pcp4   \n",
       "2     545100  RP_040825_02_D01    46815      Pgm2l1   \n",
       "3  112202624  RP_110421_01_F02    60343        Klf7   \n",
       "4        772  RP_Baylor_102947    18312        Pcp4   \n",
       "\n",
       "                                          gene-name  entrez-id chromosome  \\\n",
       "0  ATPase, Na+/K+ transporting, alpha 1 polypeptide      11928          3   \n",
       "1                           Purkinje cell protein 4      18546         16   \n",
       "2                       phosphoglucomutase 2-like 1      70974          7   \n",
       "3                Kruppel-like factor 7 (ubiquitous)      93691          1   \n",
       "4                           Purkinje cell protein 4      18546         16   \n",
       "\n",
       "  plane-of-section specimen-id fold-change target-sum contrast-sum  \\\n",
       "0         sagittal         n/a      10.208   1586.732      171.975   \n",
       "1         sagittal         n/a       8.739   1801.084      206.098   \n",
       "2         sagittal         n/a       7.103   1802.017      284.020   \n",
       "3          coronal         n/a       6.891    920.788      151.672   \n",
       "4          coronal         n/a       6.556   1130.932      194.852   \n",
       "\n",
       "   num-target-samples  num-contrast-samples  \n",
       "0                  94                   104  \n",
       "1                 106                   106  \n",
       "2                  92                   103  \n",
       "3                  74                    84  \n",
       "4                 108                   122  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results = []\n",
    "test = True\n",
    "start_row = 0\n",
    "\n",
    "# API is only able to retrieve 2000 rows at once, loop until none remain\n",
    "while test:\n",
    "    result = requests.get('%sservice::mouse_differential[set$eq\\'mouse\\'][structures1$eq\\'%s\\'][structures2$eq\\'%s\\'][start_row$eq\\'%s\\']' \\\n",
    "                          % (service, structures1, structures2, str(start_row))).json()\n",
    "    if result['success'] == True:\n",
    "        all_results += result['msg']\n",
    "        start_row += 2000\n",
    "    else:\n",
    "        test = False\n",
    "        \n",
    "# Organize into dataframe        \n",
    "mouse_differential = pd.DataFrame(all_results)\n",
    "mouse_differential.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
